### PR TITLE
feat: implement location and department management endpoints

### PR DESCRIPTION
**Pull request by eugeneanokye99**
## What
Implements Epic 3 Location & Department Management:
- US-LOC-01 / FR-LOC-01: create, list active, and update locations
- US-LOC-02 / FR-LOC-02: create and list departments within a location
- US-LOC-03 / FR-LOC-03: get locations assigned to the authenticated manager

Implemented endpoints:
- POST /api/v1/locations
- GET /api/v1/locations
- PATCH /api/v1/locations/{id}
- POST /api/v1/locations/{locationId}/departments
- GET /api/v1/locations/{locationId}/departments
- GET /api/v1/managers/me/locations

## Why
ShiftSync needs location and department structures before scheduling logic can be scoped correctly by branch/site and work area.
Managers also need visibility into only their assigned locations so authorization boundaries are explicit and enforceable.

## How to test
1. Log in as HR_ADMIN and call `POST /api/v1/locations` with:
   - `name`, `address`, `maxHeadcountPerShift`
   - Expect `201`
2. Call `POST /api/v1/locations` again with same `name`
   - Expect `409` duplicate conflict
3. Call `GET /api/v1/locations`
   - Expect only active locations
4. Call `PATCH /api/v1/locations/{id}` as HR_ADMIN to update fields
   - Expect `200` and updated payload
5. Call `PATCH /api/v1/locations/{id}` as MANAGER or EMPLOYEE
   - Expect `403`
6. As HR_ADMIN, call `POST /api/v1/locations/{locationId}/departments` with `{ "name": "Kitchen" }`
   - Expect `201`
7. Repeat same department name for same location
   - Expect `409`
8. Call `POST /api/v1/locations/{invalidLocationId}/departments`
   - Expect `404`
9. Call `GET /api/v1/locations/{locationId}/departments`
   - Expect `200` with departments for that location
10. Log in as MANAGER and call `GET /api/v1/managers/me/locations`
    - If assigned: expect list of locations
    - If not assigned: expect `200` with `[]` (empty array)
11. Call `GET /api/v1/managers/me/locations` as EMPLOYEE
    - Expect `403`

## Notes
- Intentional API versioning is used: endpoints are under `/api/v1/...` instead of unversioned `/api/...`.
- This continues the existing project-wide versioned routing approach.
- Security model remains role-based: HR_ADMIN manages locations/departments; MANAGER reads assigned locations.
- Epic 3 includes controller/service tests for location and department flows (including auth and conflict paths).

image

### CODE REVIEW

File: src/main/java/com/shiftsync/shiftsync/location/controller/ManagerLocationController.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    @Operation(
+            summary = "Get assigned locations",
+            description = "Returns locations assigned to the authenticated manager."
+    )
+    @ApiResponses(value = {
+            @ApiResponse(responseCode = "200", description = "Locations fetched", content = @Content(array = @ArraySchema(schema = @Schema(implementation = LocationResponse.class)))),
+            @ApiResponse(responseCode = "401", description = "Unauthorized", content = @Content(schema = @Schema(implementation = ErrorResponse.class))),
+            @ApiResponse(responseCode = "403", description = "Access denied", content = @Content(schema = @Schema(implementation = ErrorResponse.class)))
+    })
+    public ResponseEntity<List<LocationResponse>> getMyAssignedLocations(Authentication authentication) {
```
Comment (essilfie-b) [inline src/main/java/com/shiftsync/shiftsync/location/controller/ManagerLocationController.java:44]: **Duplicated helper — extract to a shared component (issue #2).** This exact `getCurrentUserId` method also exists in `EmployeeController` from PR #19. As more controllers are added, this will keep being copy-pasted. Extract it to an `AuthenticationHelper` Spring `@Component` and inject it wherever needed.

File: src/test/java/com/shiftsync/shiftsync/location/controller/LocationControllerWebMvcTest.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                }
+                """;
+
+        mockMvc.perform(patch("/api/v1/locations/1")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(body))
+                .andExpect(status().isForbidden());
+    }
+}
+
```
Comment (essilfie-b) [inline src/test/java/com/shiftsync/shiftsync/location/controller/LocationControllerWebMvcTest.java:190]: **Missing 400 validation test (issue #3).** This test class has no case for a missing or blank required field. Add a test that posts a body without `name` (or with `name: ""`) and asserts `status().isBadRequest()`. `CreateLocationRequest` has `@NotBlank` on `name` and `address` — verify those constraints fire correctly through the controller.

File: src/test/java/com/shiftsync/shiftsync/department/controller/DepartmentControllerWebMvcTest.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    @Test
+    @WithMockUser(roles = "EMPLOYEE")
+    void getDepartmentsByLocation_WrongRole_ReturnsForbidden() throws Exception {
+        mockMvc.perform(get("/api/v1/locations/1/departments"))
+                .andExpect(status().isForbidden());
+    }
+}
+
```
Comment (essilfie-b) [inline src/test/java/com/shiftsync/shiftsync/department/controller/DepartmentControllerWebMvcTest.java:139]: **Missing 400 validation test (issue #3).** Add a test that posts `{"name": ""}` or an empty body to `POST /api/v1/locations/1/departments` and asserts `status().isBadRequest()`. `CreateDepartmentRequest` has `@NotBlank` on `name` — exercise that constraint here.

**Review by essilfie-b — CHANGES_REQUESTED** [review]
## Overall: CHANGES REQUESTED

---

## What's good

- PR description is well-structured — What, Why, How to test, and Notes sections all present. Keep this standard.
- **Service-layer tests are now included** — `LocationServiceImplTest` and `DepartmentServiceImplTest` both cover success, duplicate, and not-found paths. This is a clear improvement from PR #19. Well done.
- All three user stories (US-LOC-01, US-LOC-02, US-LOC-03) are correctly addressed with the right endpoints and HTTP verbs.
- Case-insensitive duplicate detection (`existsByNameIgnoreCase`, `existsByLocationIdAndNameIgnoreCase`) — prevents silent duplicates like "Kitchen" vs "kitchen".
- `updateLocation` handles partial updates correctly — only mutates fields that are non-null in the request. Clean pattern.
- `existsByNameIgnoreCaseAndIdNot` prevents a location from conflicting with its own current name during an update. Exactly right.
- `getDepartmentsByLocation` returns 404 for an unknown location (per US-LOC-02) and orders results alphabetically — both correct.
- `getAssignedLocationsForManager` returns an empty list (not 404) when the manager has no assigned locations — matches US-LOC-03 exactly.
- `@Valid` applied consistently on all request bodies. Swagger/OpenAPI annotations on every endpoint.
- Controller tests (`LocationControllerWebMvcTest`, `DepartmentControllerWebMvcTest`, `ManagerLocationControllerWebMvcTest`) all cover: no token → 401, wrong role → 403, success, and conflict paths. Good breadth.

---

## Changes needed

### 1. PR title doesn't follow the git workflow format
**Ref:** `shiftsync-git-workflow.md`

The PR title is:
> `Implement location and department management endpoints`

The workflow requires commit messages and PR titles to use a prefix in imperative mood:
> `feat: implement location and department management endpoints`

**Fix:** Rename the PR title to `feat: implement location and department management endpoints`.

---

### 2. `getCurrentUserId` is duplicated across controllers
**Files:**
- `src/main/java/com/shiftsync/shiftsync/location/controller/ManagerLocationController.java` (lines 44–73)
- `src/main/java/com/shiftsync/shiftsync/employee/controller/EmployeeController.java` (from PR #19)

The exact same 30-line `getCurrentUserId` private method — handling `Long`, `String`, and `UserDetails` principal variants — is copy-pasted into both controllers. This will continue to grow as more controllers are added.

**Fix:** Extract it into a shared Spring component, for example:

```java
@Component
public class AuthenticationHelper {
    public Long getCurrentUserId(Authentication authentication) {
        // single implementation here
    }
}
```

Inject `AuthenticationHelper` into any controller that needs the current user's ID. This keeps the logic in one place and makes it testable on its own.

---

### 3. No validation-failure (400) test in controller tests
**Files:**
- `src/test/java/com/shiftsync/shiftsync/location/controller/LocationControllerWebMvcTest.java`
- `src/test/java/com/shiftsync/shiftsync/department/controller/DepartmentControllerWebMvcTest.java`

Both test classes cover auth, role, and conflict scenarios but have no test for bean validation failures. `CreateLocationRequest` has `@NotBlank` on `name` and `address`, and `@NotNull @Min(1)` on `maxHeadcountPerShift`. `CreateDepartmentRequest` has `@NotBlank` on `name`. None of these constraints are exercised in the tests.

**Fix:** Add at least one test per controller that submits a blank or missing required field and asserts `status().isBadRequest()` with field-level errors in the response body. For example:

```java
@Test
@WithMockUser(roles = "HR_ADMIN")
void createLocation_MissingName_ReturnsBadRequest() throws Exception {
    mockMvc.perform(post("/api/v1/locations")
                    .contentType(MediaType.APPLICATION_JSON)
                    .content("""
                            {
                              "address": "Airport Road",
                              "maxHeadcountPerShift": 40
                            }
                            """))
            .andExpect(status().isBadRequest());
}
```

---

## Notes

This PR represents a significant step up in quality compared to #19 — the most important feedback from that review (missing service-layer tests) was addressed here before it was even merged. That shows good awareness and discipline.

The two code changes needed before merge are the `getCurrentUserId` duplication (it will compound with every new controller) and the missing 400 validation tests (they're quick to add and close a gap in test coverage). The PR title fix is a one-liner.

Keep the momentum going.

File: src/main/java/com/shiftsync/shiftsync/location/controller/ManagerLocationController.java (hunk 1 above)
Comment (eugeneanokye99) [reply src/main/java/com/shiftsync/shiftsync/location/controller/ManagerLocationController.java:44]: I extracted the duplicated `getCurrentUserId(Authentication)` logic into a shared Spring component: `AuthenticationHelper`.

Changes made:
- Added `AuthenticationHelper` in `src/main/java/com/shiftsync/shiftsync/common/util/AuthenticationHelper.java`
- Refactored `EmployeeController` and `ManagerLocationController` to inject and use the helper
- Removed the duplicated private method from both controllers

This centralizes principal parsing in one place and avoids copy-paste as more controllers are added.

File: src/test/java/com/shiftsync/shiftsync/location/controller/LocationControllerWebMvcTest.java (hunk 2 above)
Comment (eugeneanokye99) [reply src/test/java/com/shiftsync/shiftsync/location/controller/LocationControllerWebMvcTest.java:190]: Fixed — I added a controller-level validation test for required fields in `LocationControllerWebMvcTest`.

New test:
- `createLocation_BlankName_ReturnsBadRequest()`

It posts a create payload with `name: ""` and asserts:
- `400 Bad Request`
- `fieldErrors.name = "Location name is required"`

This confirms `@NotBlank` on `CreateLocationRequest.name` is firing through the controller.

File: src/test/java/com/shiftsync/shiftsync/department/controller/DepartmentControllerWebMvcTest.java (hunk 3 above)
Comment (eugeneanokye99) [reply src/test/java/com/shiftsync/shiftsync/department/controller/DepartmentControllerWebMvcTest.java:139]: Fixed. I added the missing 400 validation test in `DepartmentControllerWebMvcTest`:

- `createDepartment_BlankName_ReturnsBadRequest()`

It posts `{"name": ""}` to `POST /api/v1/locations/1/departments` and asserts:
- `400 Bad Request`
- `fieldErrors.name = "Department name is required"`

This exercises the `@NotBlank` constraint on `CreateDepartmentRequest.name` through the controller layer.

**Review by essilfie-b — APPROVED** [review]
